### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by vlbee**

### CODE REVIEW

File: src/main/scala/Bag.scala [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    'V' -> 2,
+    'W' -> 2,
+    'Y' -> 2,
+    'X' -> 1,
+    'Z' -> 1,
+  )
+
+  def initialise(): List[Tile] = {
+    def dupeLetters(letter: (Char, Int)): List[Tile] = List.fill(letter._2)(Tile(letter._1))
+    Random.shuffle(letterDistribution.toList.flatMap(dupeLetters))
```
Comment (johnduffell) [inline src/main/scala/Bag.scala]: by having a non referentially transparent function (Random.shuffle does something different each time you call it) it makes this function non deterministic and thus not fully unit testable.

A common technique is to push any non determinism as high up in the code as possible, so ideally the top level function would just be calling the deterministic main part of the program, and then running all the side effects/impure parts.

In this case that would be to always return the list in order, and shuffle it in the parent function.

**Comment by JonNorman** [comment]
Ah great, I had a branch that added loads of comments, so I'll move them to this PR instead!

File: src/main/scala/Bag.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    'U' -> 4,
+    'V' -> 2,
+    'W' -> 2,
+    'Y' -> 2,
+    'X' -> 1,
+    'Z' -> 1,
+  )
+
+  def initialise(): List[Tile] = {
+    def dupeLetters(letter: (Char, Int)): List[Tile] = List.fill(letter._2)(Tile(letter._1))
```
Comment (JonNorman) [inline src/main/scala/Bag.scala]: Nice use of List.fill! You could express this in a slightly more idiomatic way:

```scala
Random.shuffle(
    letterDistribution.flatMap {
        case (letter, distribution) => List.fill(distribution)(Tile(letter))
     }.toList
)
```

File: src/main/scala/Game.scala [hunk 3]
```diff
@@ -0,0 +1,6 @@
+case class Game (
```
Comment (JonNorman) [inline src/main/scala/Game.scala]: You might want to create a package for your various models - e.g. `scrabble.model` - and stick `Game`, `Player`, `Tile` and `Bag` in it as an easy way to move around them. Or, if your models don't have too much extra logic around them then, you could just have a file called `Models` and stick them all in there.

File: src/main/scala/GameHelpers.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import scala.util.Random
+
+object GameHelpers {
+
+  def fillRack(rack: List[Tile], letterBag: List[Tile]): (List[Tile], List[Tile]) = {
+    val (drawnTiles, newBag) = letterBag.splitAt(7 - rack.length)
+    val filledRack = rack ++ drawnTiles
+
+    if (filledRack.length < 7) {
```
Comment (JonNorman) [inline src/main/scala/GameHelpers.scala]: Why not check if the `letterBag.isEmpty`?

File: src/main/scala/GameHelpers.scala [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    val wordTiles: List[Tile] = word.toList.map(c => Tile(c))
+    rack diff wordTiles
+  }
+
+  def calculateWordScore(word: String): Int = {
+    word.toList.map(Tile(_).value).sum
+  }
+
+  def isInDictionary(word: String): Boolean = {
+    if (Dictionary.all.contains(word.toLowerCase())) true else false
```
Comment (JonNorman) [inline src/main/scala/GameHelpers.scala]: Any time you see a statement returning `true` or `false` ask whether you need to use these literals. What would happen if you just returned
```scala
Dictionary.all.contains(word.toLowerCase)
```
?

File: src/main/scala/GameHelpers.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  def isInRack(word: String, rack: List[Tile]): Boolean = {
+    val sortedWord = word.toList.sortWith(_.compareTo(_) < 0)
```
Comment (JonNorman) [inline src/main/scala/GameHelpers.scala]: I think you can do
```scala
word.toList.sorted(_.toString)
```

As `sorted` needs something that implements `Ordered` and `String` does, so by converting to a `String` we let the ordering take care of itself.

File: src/main/scala/GameHelpers.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    val sortedRack = rack.map(_.letter).sortWith(_.compareTo(_) < 0)
```
Comment (JonNorman) [inline src/main/scala/GameHelpers.scala]: and here

File: src/main/scala/GameHelpers.scala [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    def loop(wL: List[Char], wI: Int, rL: List[Char], rI: Int): Boolean = {
```
Comment (JonNorman) [inline src/main/scala/GameHelpers.scala]: It's great that you understand recursion enough to write this! Bravo! Another way to write this might be:

```scala
import scala.annotation.tailrec

@tailrec
def isInRack(word: String, rack: List[Tile]): Boolean = {
    go(word.toList, rack.map(_.c))

    @tailrec
    def go(wordLetters: List[Char], rackLetters: List[Char]): Boolean =
        wordLetters match {
            case Nil => true
            case h :: t => rackLetters.contains(h) && go(t, rackLetters diff List(h))
        }
}
```

File: src/main/scala/GameHelpers.scala [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    loop(sortedWord, 0, sortedRack, 0)
+  }
+
+  def isValidWord(word: String, rack: List[Tile]): Boolean = {
+    if(!isInDictionary(word)) { println(s"'$word' is not in the Scrabble Dictionary.")}
+    if(!isInRack(word, rack)) { println("You don't have those letter tiles in your rack.")}
+    isInDictionary(word) && isInRack(word, rack)
+  }
+
+  def playWord(rack: List[Tile], level: String): Option[String] = {
```
Comment (JonNorman) [inline src/main/scala/GameHelpers.scala]: If this is only for computer players - you might consider renaming this to make it more explicit.

File: src/main/scala/GameHelpers.scala [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    val possibleWords = Dictionary.all.filter(word => isInRack(word.toUpperCase, rack))
+    if (possibleWords == Nil) {
+      None
+    } else {
+      level match {
+        case "easy" => Some(Random.shuffle(possibleWords).head.toUpperCase)
+        case "hard" => Some(possibleWords.sortBy(word => (word.length)).reverse.head.toUpperCase)
```
Comment (JonNorman) [inline src/main/scala/GameHelpers.scala]: Nice! There are lots of ways you could go about this e.g. pick the word with the biggest scoring value.

File: src/main/scala/GamePrompts.scala [hunk 11]
```diff
@@ -0,0 +1,49 @@
+import GameHelpers.{calculateWordScore}
+
+object GamePrompts {
+
+  def promptName(): String = {
+    println("\u001b[33;1mWhat is your name?")
```
Comment (JonNorman) [inline src/main/scala/GamePrompts.scala]: No closing tag here means that the colour code persists when the user provides their input. Looks pretty, but not sure if you intended that or not!

File: src/main/scala/GamePrompts.scala [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    scala.io.StdIn.readLine().toLowerCase.capitalize
+  }
+
+  def promptLevel(): String = {
```
Comment (JonNorman) [inline src/main/scala/GamePrompts.scala]: It looks like there is a pattern of prompting with the `\u001b` colour and potentially asking for a response. Could we factor this out anywhere and avoid some of this repetition? Would it be worth it?
```scala
def notifyUser(message: String): Unit =
    println(s"\u001b[33;1m$question\u001b[0m ")

def promptUser(question: String): String = {
    notifyUser(question)
    scala.io.StdIn.readLine().toString
}
```

File: src/main/scala/GamePrompts.scala [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    println("\u001b[33;1mHow good are you at Scrabble?\u001b[0m ")
+    println("[1] I'm a pro.")
+    println("[2] Go easy.")
+    println()
+    val level = scala.io.StdIn.readLine().toString
```
Comment (JonNorman) [inline src/main/scala/GamePrompts.scala]: Nice touch adding a difficulty 👍

File: src/main/scala/GameRunner.scala [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    println("Welcome to \u001b[31;1mScala Scrabble Practice")
+    println()
+    val name: String = promptName()
+    println()
+    println("Hello " + name + ". \u001b[0mThe Human plays against the Computer.")
+    println()
+    val level: String = promptLevel()
+
+    val letterBag = Bag.initialise()
+    val human = Player(name, Nil, Nil, false)
```
Comment (JonNorman) [inline src/main/scala/GameRunner.scala]: In the same way that you have a `Bag.initialise`, why not create one for `Player` to save yourself some boilerplate? Perhaps

```scala
def apply(???): Player = ???
```

File: src/main/scala/GameRunner.scala [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    val computer = Player("The Computer", Nil, Nil, false)
+    val newGame = Game(letterBag, human, computer, level)
```
Comment (JonNorman) [inline src/main/scala/GameRunner.scala]: Nice encapsulation of all the state required to run the game 👍! How would you model this if we wanted to have a 3-player game of Scrabble? Or 6 player?

File: src/main/scala/GameRunner.scala [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+    humanTurn(newGame)
```
Comment (JonNorman) [inline src/main/scala/GameRunner.scala]: This logic of each `xTurn` calling the next one a bit tricky. How about a loop that checks whether the `Game` is over and has a pointer to whose turn is next? Then none of the `xTurn` methods need to know who else they are playing, they just take their turn when prompted and then wait for their go again. This would make it simpler to reason about as the `Player` model has fewer concerns.

In addition, would it be possible to model the distinction between a Human and Computer player directly? Then your game interface could be much simpler, such as 

```scala
def takeTurn(game: Game, player: Player): Game
```

If you can get to the point where you've abstracted away the differences between a Human and a Computer, you can write a single `takeTurn` method and everything will fall into place. This would then also allow you to only write some of the logic once e.g. calls to `checkGameOver`, `displayRack`, `word match etc.

File: src/main/scala/GameRunner.scala [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+    val computer = game.computer
+    val (rack, bag) = fillRack(computer.rack, game.bag)
+    displayRack(computer.name, rack)
+
+    val word = playWord(rack, game.level)
+
+    word match {
+      case Some(w) =>
+        println(computer.name + s" has played: '$w'")
```
Comment (JonNorman) [inline src/main/scala/GameRunner.scala]: If you're using string interpolation here, then you can avoid doing "" + "" and put it all in one string 👍

File: src/main/scala/GameRunner.scala [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+    checkGameOver(game)
+
+    val human = game.human
+    val (rack, bag) = fillRack(human.rack, game.bag)
+    displayRack(human.name, rack)
+
+    val word: String = promptWord()
+
+    if (bag == Nil && word == "") {
+      println("You forfeited but as the letter bag is empty, it's \u001b[33;1mthe end of the game for you.")
```
Comment (JonNorman) [inline src/main/scala/GameRunner.scala]: You? YOU?!

You've got my name! Please use it! 😭

File: src/main/scala/GameRunner.scala [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+      val playedRack = removeTilesFromRack(word, rack)
+      displayScore(human, word)
+      val nextHuman = Player(human.name, playedRack, human.wordsPlayed ++ List(word), false)
+      computerTurn(Game(bag, nextHuman, game.computer, game.level))
+    }
+  }
+
+  def checkGameOver(game: Game): Unit = {
+    val name = game.human.name
+    val computerScore: Int = (game.computer.wordsPlayed).map(calculateWordScore(_)).sum
```
Comment (JonNorman) [inline src/main/scala/GameRunner.scala]: If the calculation of score depends exclusively on the other attributes of a given `Player` then I would consider adding a helper method to either the `Player` case class or creating a `Player` object and adding it there as it would stop us repeating ourselves here and can then easily be tested.

File: src/main/scala/GameRunner.scala [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    val humanScore: Int = (game.human.wordsPlayed).map(calculateWordScore(_)).sum
+
+    if (game.bag == Nil && (game.human.gameOver || game.computer.gameOver)) {
```
Comment (JonNorman) [inline src/main/scala/GameRunner.scala]: `game.bag.isEmpty` might be more idiomatic here or (as above with `Player.calculateScore`) add a helper method

```scala
def isGameOver(game: Game): Boolean = ???
```

to make this clearer and more explicit.

File: src/main/scala/GameRunner.scala [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+      println("\u001b[31;1mIt's GAME OVER.")
```
Comment (JonNorman) [inline src/main/scala/GameRunner.scala]: It's nice to have small functions that do one thing well, how about moving the text below to a
```scala
def printEndGameSummary(game: Game): Unit
```

and then calling that here?

File: src/main/scala/Player.scala [hunk 22]
```diff
@@ -0,0 +1,6 @@
+case class Player(
+                   name: String,
+                   rack: List[Tile],
+                   wordsPlayed: List[String],
+                   gameOver: Boolean
```
Comment (JonNorman) [inline src/main/scala/Player.scala]: If we were playing scrabble would we normally think about a Player's game being over, or the whole game being over? Does it make sense to model this at the `Player` level?

File: src/main/scala/Tile.scala [hunk 23]
```diff
@@ -0,0 +1,18 @@
+case class Tile(c: Char) {
+
+  def value: Int = {
```
Comment (JonNorman) [inline src/main/scala/Tile.scala]: This is an interesting case! What if someone puts in a word like "groß"? Presumably it wouldn't be allowed because the tiles aren't in the bag, so it would fail further up in the validation, but it would be nice not to have to be concerned about that here, if we could just use local reasoning.

If, instead of matching a `Char`, we were to ensure that a `Tile` can only be one of A-Z, then here we would be able to have an exhaustive check and not worry about the `_` case. Better yet, we could fix the value of a given Tile/Letter so that we don't have to match at all:

```scala
sealed trait Tile {
    val value: Int
}
object Tile {
    case object A extends Tile { override val value = 1 }
    case object B extends Tile { override val value = 3 }
}
```

File: src/main/scala/Tile.scala [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+      case 'B' | 'C' | 'M' | 'P' => 3
+      case 'F' | 'H' | 'V' | 'W' | 'Y' => 4
+      case 'K' => 5
+      case 'J' | 'X' => 8
+      case 'Q' | 'Z' => 10
+      case _ => 0
+    }
+  }
+
+  def letter: Char = c
```
Comment (JonNorman) [inline src/main/scala/Tile.scala]: What's the reason for this alias? Can we just rename `c` to `letter`?

**Review by JonNorman — COMMENTED** [review]
Really nice work! A load of my comments are food for thought, nit picks or suggestions. Great effort so far 👍 🙌

File: src/main/scala/Bag.scala (hunk 2 above)
Comment (johnduffell) [reply src/main/scala/Bag.scala]: agree although `distribution` should be `quantity`

File: src/main/scala/Dictionary.scala [hunk 25]
```diff
@@ -0,0 +1,5 @@
+import scala.io.Source
+
+object Dictionary {
+  val all: List[String] = Source.fromFile("src/main/resources/dictionary.txt").getLines.toList
```
Comment (johnduffell) [inline src/main/scala/Dictionary.scala]: for java as well as scala this would only work if you are running locally as the source files happen to be under the root of the project.
If you are building the code into a deplyable JAR file then the dictionary would be on the classpath but not in the file system.
You can get at instead with something like 
```scala
Source.fromInputStream(getClass.getResourceAsStream("dictionary.txt"))
```

File: src/main/scala/GamePrompts.scala [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+    level match {
+      case "1" => "hard"
+      case "2" => "easy"
+      case _ => promptLevel()
```
Comment (johnduffell) [inline src/main/scala/GamePrompts.scala]: rather than using a String, you should use an algebraic data type - i.e.
```scala
sealed abstract class Level(val label: String)
case object Easy extends Level("easy")
case object Hard extends Level("hard")
```
Then you can pass Level instead of String, which also means that when you pattern match  in playWord, the compiler will check that you didn't miss any or mis spell anything.  And it also makes sure it didn't mix with any other strings.

File: src/main/scala/GameHelpers.scala [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+    if(!isInRack(word, rack)) { println("You don't have those letter tiles in your rack.")}
+    isInDictionary(word) && isInRack(word, rack)
+  }
+
+  def playWord(rack: List[Tile], level: String): Option[String] = {
+    val possibleWords = Dictionary.all.filter(word => isInRack(word.toUpperCase, rack))
+    if (possibleWords == Nil) {
+      None
+    } else {
+      level match {
```
Comment (johnduffell) [inline src/main/scala/GameHelpers.scala]: you could do something along these lines to split out the common parts which keeping the structure a bit flatter
```scala
Some(possibleWords).filter(_.nonEmpty).map { possibleWords => 
  level match {
    case "easy" => Random.shuffle(possibleWords)
    case "hard" => possibleWords.sortBy(word => (word.length)).reverse
  }
}.map(_.head.toUpperCase)
```

File: src/main/scala/GameHelpers.scala [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+        case "easy" => Some(Random.shuffle(possibleWords).head.toUpperCase)
+        case "hard" => Some(possibleWords.sortBy(word => (word.length)).reverse.head.toUpperCase)
+      }
+    }
+  }
+
+  def flagPlayerGameOver(bag: List[Tile]): Boolean = {
+    if (bag == Nil) true else false
```
Comment (johnduffell) [inline src/main/scala/GameHelpers.scala]: just change the whole line to 
```suggestion
    bag.isEmpty
```

File: src/main/scala/GameHelpers.scala [hunk 29]
```diff
@@ -0,0 +1,69 @@
+import scala.util.Random
+
+object GameHelpers {
```
Comment (johnduffell) [inline src/main/scala/GameHelpers.scala]: I avoid using non specific words such as "Helpers" as all code should be helping with something specific.  It can turn into a bit of a catch all otherwise.  A good name should exclude more stuff than it includes.  The same applies for "common" and "service".
It does make naming a lot harder, because you have to keep thinking of specific names, but it helps you focus on the higher level structure a bit.

File: src/main/scala/Player.scala [hunk 30]
```diff
@@ -0,0 +1,6 @@
+case class Player(
+                   name: String,
```
Comment (johnduffell) [inline src/main/scala/Player.scala]: for indendation I would recommend to stick to 2 space everywhere, column formatting of code often ends with lots of lines needing reformatting for a simple change elsewhere e.g. rename the player class to something longer/shorter means you have to reindent all the lines.

File: src/main/scala/Tile.scala [hunk 31]
```diff
@@ -0,0 +1,18 @@
...
+  def value: Int = {
+    c match {
+      case 'A' | 'E' | 'I' | 'O' | 'U' | 'N' | 'R' | 'T' | 'L' | 'S' => 1
+      case 'G' | 'D' => 2
+      case 'B' | 'C' | 'M' | 'P' => 3
+      case 'F' | 'H' | 'V' | 'W' | 'Y' => 4
+      case 'K' => 5
+      case 'J' | 'X' => 8
+      case 'Q' | 'Z' => 10
+      case _ => 0
```
Comment (johnduffell) [inline src/main/scala/Tile.scala]: this case is only needed to keep the compiler happy, but actually we know that this case would be a program bug.  You could use a type that only had valid letters, you'd have a hassle to model it ( a sealed trait or sealed class with 26 different subclasses) but once you had modelled it, the type checker would be able to help out more in the logic parts.
That is part of the classic trade off of the more you help the compiler out, the more it helps you, however you can end up making a lot of code just to put the types in in the first place.

File: src/test/scala/BagSpec.scala [hunk 32]
```diff
@@ -0,0 +1,10 @@
+import org.scalatest._
+
+class BagSpec extends FlatSpec with Matchers {
+
+  "initialize()" should "generate a random List of 101 Tiles" in {
+    Bag.initialise() should have length (101)
+    // TODO test for randomness
```
Comment (johnduffell) [inline src/test/scala/BagSpec.scala]: I don't think you need to handle randomness in your test, see my comment about pushing the non determinism to the edge.  One approach is to pass in a "randomise" function, and then check that it's been called correctly.  But in the real code you would just pass in the actual Random.shuffle wheras in the test you can pass in a test function.

File: src/test/scala/TileSpec.scala [hunk 33]
```diff
@@ -0,0 +1,11 @@
...
+
+class TileSpec extends FlatSpec with Matchers {
+  it should "return correct tile letter value" in {
+    Tile('A').value should be(1)
+    Tile('Z').value should be(10)
+    Tile('C').value should be(3)
+    Tile('G').value should be(2)
+    Tile('K').value should be(5)
+  }
+}
```
Comment (johnduffell) [inline src/test/scala/TileSpec.scala]: you can use some code coverage runners in IntelliJ  and that will help you understand which code is not covered if any.  See the run with coverage in the menu.
image


They also can run in your CI build. I have used https://coveralls.io which is relatively easy to set up with https://travis-ci.org
This might all be

Comment (JonNorman) [reply src/test/scala/TileSpec.scala]: Might be a good candidate for using scala test generators as well.

File: src/main/scala/Bag.scala (hunk 1 above)
Comment (vlbee) [reply src/main/scala/Bag.scala]: Makes sense! Could complete the test and also learnt that Maps do not preserver order!

File: src/main/scala/GameHelpers.scala (hunk 6 above)
Comment (vlbee) [reply src/main/scala/GameHelpers.scala]: That did not work, but `word.toList.sorted` did :)

File: src/main/scala/GameHelpers.scala (hunk 27 above)
Comment (vlbee) [reply src/main/scala/GameHelpers.scala]: 🤔  not sure I understand what `Some(possibleWords).filter(_.nonEmpty).map` is doing

Comment (johnduffell) [reply src/main/scala/GameHelpers.scala]: so to answer your specific question it's returning `None` if the possible words is empty otherwise it's returning a `Some` with whatever the output of the function inside the map is.

In real terms I just noticed your function is returning an Option, so rather than starting with if statements and spreading the Some/None around, it places the option up front and then processes the thing that's inside it (we don't have to keep mentioning Some or None after that point).

The only benefit is this is a way of restricting what the code can do to aid predictability and testability - as soon as you see "map" you know that the result must remain a "Some" without having to look inside or test the code inside.  The same goes for "filter" you know the only thing it can do is convert to None.

Having said all that, this is a typical functional programming way of doing abstraction and it can be less descriptive, so that's why I wrote you "could" do that rather than you "should"!  Grab me if you want any clarification!
